File: lua/binary-search/binary-search.lua
return function(array, target)
  local floor = math.floor
  local ceil = math.ceil
  local found = false

  local index = #array == 1 and 1 or floor(#array / 2)
  print("target:"..target)
  local elements = #array
  while elements > 0 do
    local value = array[index]
    print("index: ", index, " value: ", value, " elements: ", elements)
    if value == nil then
      elements = 0
      break
    end

    if target == value then
      found = true
      elements = 0
      break
    end

    if target < value then
      -- target = 2
      -- local array = TracedArray{ 6, 67, 123, 345, 456, 457, 490, 2002, 54321, 54322 }
      if index == 1 then
        elements = 0
        break
      end
      elements = #array - index
      index = ceil(index / 2)
    else
      -- print("else", index)
      elements = #array - index
      index = index + ceil((#array - index) / 2)
    end

  end
  if found then
    return index
  else
    return -1
  end
end

File: clojure/binary-search/src/binary_search.clj
(ns binary-search)

(defn middle [coll]
  (int (Math/floor (/ (count coll) 2))))

(defn fff [v]
  (prn "v" v)
  (let [v-copy v]
    (prn "KEYS:" (keys v-copy))
    (recur (rest v))))

(defn search-for [n coll]
  (prn "search-for n" n)
  (prn "search-for coll" coll)
  (let [cells (vec (map-indexed vector coll))]
    (prn "cells" cells)
    (loop [n n c cells]
      (prn "n" n)
      (prn "c" c)
      (let [m (middle c) [i x] (get c m)]
        (prn "m" m)
        (prn "i" i)
        (prn "x" x)
        (cond
          (and (not= n x) (= 1 (count c))) (throw (Exception. "not found"))
          (> n x) (do (prn "> n x") (recur n (vec (nthrest c m))))
          (< n x) (do (prn "< n x") (recur n (vec (take (inc m) c))))
          :else (do (prn "else") i))))))
  
(comment

  (as-> [[0 1] [1 3] [2 5] [3 8] [4 13] [5 21] [6 34] [7 55] [8 89]] c
    (vec (take 4 c)))

  (as-> [[0 1] [1 3] [2 4] [3 6] [4 8] [5 9] [6 11]] c
       (let [mid (middle c)]
         (get c mid)
         (vec (nthrest c mid))))

  (map-indexed vector [77 88 99 111 222])
  (defn search-for [n coll]
    (let [m (middle coll) x (get coll m)]
      (prn "m" m)
      (prn "x" x)
      (->> (map vector (iterate inc 1))
        (cond nthrest))))

  (defn search-for [n coll]
    (prn "n" n)
    (prn "coll" coll)
    (let [m (middle coll) x (get coll m)]
      (prn "m" m)
      (prn "x" x
      ; (if (= n x)
      ;   m
        (cond
          (= n x) (do
                      (prn "= n x")
                      m)
          (> n x) (do
                    (prn "> n x")
                    (search-for n (vec (nthrest coll m))))
          (< n x) (do
                    (prn "< n x")
                    (let [newcoll (map fff coll)]
                      (prn "newcoll" newcoll)
                      (search-for n newcoll))))))))
